### DISCUSSION TITLE
Remove/change event not firing

### DISCUSSION THREAD

**Discussion by psaju** [clientIO/joint#2733] [Q&A]
### Introduction

I'm writing an app that generates wiring diagrams. Every fusebox is a different diagram, and the user has the possibility to add comments to the diagram. Switching between the diagrams causes the graph to be replaced with another one. The issue I'm having is that no change or remove events are fired for the diagrams (add works ok), which causes the diagram not to update, even if the view is correct.

I have added the code as an attachment in the zip file.

[diagramService.zip](https://github.com/user-attachments/files/16560479/diagramService.zip)

[CodeSandBox](https://codesandbox.io/p/sandbox/joint-js-diagram-service-xp8p6j?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clzmmjugj0006356l0oeo2ycn%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clzmmjugj0002356lfyd5hhtt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clzmmjugj0003356lli7ahz9c%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clzmmjugj0005356lj6giog93%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clzmmjugj0002356lfyd5hhtt%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzmmjugj0001356lerb08ixc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.html%2522%257D%255D%252C%2522id%2522%253A%2522clzmmjugj0002356lfyd5hhtt%2522%252C%2522activeTabId%2522%253A%2522clzmmjugj0001356lerb08ixc%2522%257D%252C%2522clzmmjugj0005356lj6giog93%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzmmjugj0004356ldl3oz9s8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%255D%252C%2522id%2522%253A%2522clzmmjugj0005356lj6giog93%2522%252C%2522activeTabId%2522%253A%2522clzmmjugj0004356ldl3oz9s8%2522%257D%252C%2522clzmmjugj0003356lli7ahz9c%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clzmmjugj0003356lli7ahz9c%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
How to reproduce the problem? I'm not sure what to look at in the file.
Note that `remove` event is not triggered if you call `resetCells()` or `graph.fromJSON()`. A single `reset` event is fired instead.

**Reply by psaju** [reply]
In initGraph I initialize the paper and graph. In generateDiagramData I generate graphs off screen for them to be ready upon loading the DOM. In renderDiagram in render the currently selected graph. In initEvents I initialize different listeners.

The problem I am having is that the graphs aren't always updated - I can dynamically add a cell to a diagram and it persists after switching the graphs. If I change the added cell's label or position it does not persist. I want to save the data, and the obvious way would be to add a graph.on('change') listener and save the data in the callback, but that listener does not fire, but graph.on('all') does fire on the change event. I'm thinking that maybe I made some kind of logical error or false assumption about the graph should work, and that may cause further problems down the road.

**Reply by kumilingus** [reply]
Are you able to create a minimal example that demonstrates the issue?
```ts
const g1 = new dia.Graph();
const g2 = new dia.Graph();
/* ... */
g1.on('change', console.log);
/* .. */
cell.set('attribute', 'value');

// no change event is triggered
```
